File: day11/day11.cpp
#include "utils.h"


using namespace std;


bool step1(vector<string>& grid1, vector<string>& grid2)
{
    for (int r = 1; r < (grid1.size() - 1); ++r)
    {
        for (int c = 1; c < (grid1[0].size() - 1); ++c)
        {
            int occ = 0;
            //occ += (grid1[r  ][c] == '#') ? 1 : 0;
            occ += (grid1[r+1][c]   == '#') ? 1 : 0;
            occ += (grid1[r-1][c]   == '#') ? 1 : 0;
            occ += (grid1[r  ][c+1] == '#') ? 1 : 0;
            occ += (grid1[r+1][c+1] == '#') ? 1 : 0;
            occ += (grid1[r-1][c+1] == '#') ? 1 : 0;
            occ += (grid1[r  ][c-1] == '#') ? 1 : 0;
            occ += (grid1[r+1][c-1] == '#') ? 1 : 0;
            occ += (grid1[r-1][c-1] == '#') ? 1 : 0;

            if ((occ == 0) && (grid1[r][c] == 'L'))
                grid2[r][c] = '#';
            else if ((occ >= 4) && (grid1[r][c] == '#'))
                grid2[r][c] = 'L';
            else 
                grid2[r][c] = grid1[r][c];
        }
    }

    return (grid2 == grid1);
}


int visible(vector<string>& grid1, int r, int c, int rdel, int cdel)
{
    while (true)
    {
        r += rdel;
        if ((r == 0) || (r == (grid1.size() - 1)))
            return 0; 
        c += cdel;
        if ((c == 0) || (c == (grid1[0].size() - 1)))
            return 0;
        if (grid1[r][c] == '#')
            return 1;
        if (grid1[r][c] == 'L')
            return 0;
    }
    return 0;
}


bool step2(vector<string>& grid1, vector<string>& grid2)
{
    for (int r = 1; r < (grid1.size() - 1); ++r)
    {
        for (int c = 1; c < (grid1[0].size() - 1); ++c)
        {
            int occ = 0;
            occ += visible(grid1, r, c,  1,  1);
            occ += visible(grid1, r, c,  1,  0);
            occ += visible(grid1, r, c,  1, -1);
            occ += visible(grid1, r, c,  0,  1);
            //occ += visible(grid1, r, c,  0,  0);
            occ += visible(grid1, r, c,  0, -1);
            occ += visible(grid1, r, c, -1,  1);
            occ += visible(grid1, r, c, -1,  0);
            occ += visible(grid1, r, c, -1, -1);

            if ((occ == 0) && (grid1[r][c] == 'L'))
                grid2[r][c] = '#';
            else if ((occ >= 5) && (grid1[r][c] == '#'))
                grid2[r][c] = 'L';
            else 
                grid2[r][c] = grid1[r][c];
        }
    }

    return (grid2 == grid1);
}


int part1(const vector<string>& grid)
{
    vector<string> grid2 = grid;
    vector<string> grid1 = grid;
    while (true)
    {
        if (step1(grid1, grid2)) break;
        if (step1(grid2, grid1)) break;
    }        

    int result = 0;
    for (int r = 1; r < grid1.size() - 1; ++r)
        for (int c = 1; c < grid1[0].size() - 1; ++c)
            if (grid1[r][c] == '#')
                result += 1;
    return result;
}


// Not exactly DRY, but not important for AOC...
int part2(const vector<string>& grid)
{
    vector<string> grid2 = grid;
    vector<string> grid1 = grid;
    while (true)
    {
        if (step2(grid1, grid2)) break;
        if (step2(grid2, grid1)) break;
    }        

    int result = 0;
    for (int r = 1; r < grid1.size() - 1; ++r)
        for (int c = 1; c < grid1[0].size() - 1; ++c)
            if (grid1[r][c] == '#')
                result += 1;
    return result;
}


void solution(const char* file)
{
    cout << "Input: " << file << '\n';
    std::ifstream is{file};
    auto lines = aoc::read_lines(is);

    // Create a grid with a border to avoid edge checks.
    vector<string> grid;
    grid.push_back(string(lines[0].size() + 2, '.'));
    for (const auto& line: lines)
    {
        grid.push_back(string(string{'.'} + line + '.'));
    }
    grid.push_back(string(lines[0].size() + 2, '.'));

    auto p1 = part1(grid);
    cout << "Part1: " << p1 << '\n';

    auto p2 = part2(grid);
    cout << "Part2: " << p2 << '\n';
}


int main(int argc, char** argv)
{
    solution("day11-test.txt");
    solution("day11-input.txt");
}

File: day11/day11.py
import re
import sys
import os
sys.path.insert(1, os.path.join(sys.path[0], '..'))
import utils.utils as ut


def pack1(rows):
    rows2 = []
    for row in rows:
        rows2.append(row[:])

    for r in range(1, len(rows) - 1):
        for c in range(1, len(rows[0]) - 1):
            cells = [x for x in rows[r][c-1:c+2]] + [x for x in rows[r-1][c-1:c+2]] + [x for x in rows[r+1][c-1:c+2]]
            occ   = len([x for x in cells if x == '#'])
            if (occ == 0) and (rows[r][c] == 'L'):
                rows2[r][c] = '#'
            elif (occ >= 5) and (rows[r][c] == '#'):
                rows2[r][c] = 'L'

    return rows2


def see(rows, r, c, rdel, cdel):
    while True:
        r += rdel
        if r == 0 or r == (len(rows) - 1):
            return 0 
        c += cdel
        if c == 0 or c == (len(rows[0]) - 1):
            return 0 
        if rows[r][c] == '#':
            return 1
        if rows[r][c] == 'L':
            return 0
    return 0


def pack2(rows):
    rows2 = []
    for row in rows:
        rows2.append(row[:])

    for r in range(1, len(rows) - 1):
        for c in range(1, len(rows[0]) - 1):
            occ  = 0
            occ += see(rows, r, c,  1, 1)
            occ += see(rows, r, c,  0, 1)
            occ += see(rows, r, c, -1, 1)
            occ += see(rows, r, c,  1, 0)
            #occ += see(rows, r, c,  0, 0)
            occ += see(rows, r, c, -1, 0)
            occ += see(rows, r, c,  1, -1)
            occ += see(rows, r, c,  0, -1)
            occ += see(rows, r, c, -1, -1)

            if (occ == 0) and (rows[r][c] == 'L'):
                rows2[r][c] = '#'
            elif (occ >= 5) and (rows[r][c] == '#'):
                rows2[r][c] = 'L'

    return rows2


def changes(rows, rows2):
    for r in range(1, len(rows) - 1):
        for c in range(1, len(rows[0]) - 1):
            if rows[r][c] != rows2[r][c]:
                return True
    return False


def occupied(rows):
    occ = 0
    for r in range(1, len(rows) - 1):
        for c in range(1, len(rows[0]) - 1):
            if rows[r][c] == '#':
                occ += 1
    return occ


def part1(rows):    
    rows2 = pack1(rows)
    while changes(rows, rows2) == True:
        rows = rows2
        rows2 = pack1(rows)
    return occupied(rows2)


def part2(rows): 
    rows2 = pack2(rows)
    while changes(rows, rows2) == True:
        rows = rows2
        rows2 = pack2(rows)
    return occupied(rows2)


def solution(file):
    print("Input: ", file)

    lines = ut.read_lines(file)
    rows = []
    rows.append(list('.' * (len(lines[0]) + 2)))
    for line in lines:
        rows.append(list('.' + line + '.'))
    rows.append(list('.' * (len(lines[0]) + 2)))

    p1 = part1(rows)
    print("Part1: ", p1)

    p2 = part2(rows)
    print("Part2: ", p2)


def main():
    solution("day11-test.txt");
    solution("day11-input.txt");


if __name__ == '__main__':
    main()

